### PR TITLE
[#1959] [WebUI] Allow user selectable GUI language

### PR DESCRIPTION
**Pull request by bendikro**
Implement [#1959](http://dev.deluge-torrent.org/ticket/1959) Allow user selectable GUI language

### CODE REVIEW

File: deluge/core/daemon_entry.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import deluge.configmanager
 import deluge.error
 from deluge.ui.baseargparser import BaseArgParser
+from deluge.ui.util import lang
```
Comment (cas--) [inline deluge/core/daemon_entry.py]: I was hoping to not have a dependency on `ui` in core...

File: deluge/ui/util/lang.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+import deluge.common
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())  # Silence: No handlers could be found for logger "deluge.util.lang"
+
+
+def set_dummy_trans():
+    import __builtin__
+    __builtin__.__dict__["_"] = lambda x: x
```
Comment (cas--) [inline deluge/ui/util/lang.py]: should add a log warning that translation is unavailable but strings have been marked for translation

**Comment by cas--** [comment]
A few notes:
- English is not an option in the drop-down.
- It seems to require restarting the web-server to change language, is this unavoidable?
- Need to hit `apply` then `ok` but this is probably a pre-existing issue
- Change `height: 22` to `height: 17` to gain some height in Interface.js

**Comment by cas--** [comment]
With the extra height I would like `language` to be a `xtype: fieldset`

File: deluge/ui/web/server.py [hunk 3]
```diff
@@ -7,6 +7,8 @@
 # See LICENSE for more details.
 #
 
+from __future__ import print_function
```
Comment (bendikro) [inline deluge/ui/web/server.py]: Oh, yeah

File: deluge/core/daemon_entry.py (hunk 1 above)
Comment (bendikro) [inline deluge/core/daemon_entry.py]: It only needs this:
```python
def set_dummy_trans():                                                                                                                                                                      
    import __builtin__                                                                                                                                                                      
    __builtin__.__dict__["_"] = lambda x: x                                                                                                                                                 
```

Comment (cas--) [inline deluge/core/daemon_entry.py]: I can't see any other use for it in `lang.py` so perhaps move it here is a good idea

File: deluge/ui/util/lang.py (hunk 2 above)
Comment (bendikro) [inline deluge/ui/util/lang.py]: Log warning where? in `set_dummy_trans()` or when `_` is called?

**Comment by bendikro** [comment]
Haven't looked at the apply issue, but the Language option should now be working 100%.

File: deluge/ui/util/lang.py (hunk 2 above)
Comment (cas--) [inline deluge/ui/util/lang.py]: When `_` is called

File: deluge/core/daemon_entry.py (hunk 1 above)
Comment (bendikro) [inline deluge/core/daemon_entry.py]: It's called from setup_translations as well

**Comment by cas--** [comment]
Should we perhaps either prompt 'refresh browser' or actually refresh the page after clicking apply?

File: deluge/ui/util/lang.py (hunk 2 above)
Comment (bendikro) [inline deluge/ui/util/lang.py]: Done

**Comment by bendikro** [comment]
Probably a good idea

File: deluge/core/daemon_entry.py [hunk 4]
```diff
@@ -55,7 +56,7 @@ def start_daemon(skip_start=False):
         deluge.core.daemon.Daemon: A new daemon object
 
     """
-    deluge.common.setup_translations()
+    lang.set_dummy_trans()
```
Comment (cas--) [inline deluge/core/daemon_entry.py]: Was meant to set `warn_msg=True`?

Comment (bendikro) [inline deluge/core/daemon_entry.py]: Ah, yeah

**Comment by cas--** [comment]
You might be right that auto-refresh could be intrusive so popup box perhaps?

**Comment by bendikro** [comment]
Yeah, if that's a simple fix?

**Comment by cas--** [comment]
It should be with: https://docs.sencha.com/extjs/3.4.0/#!/api/Ext.MessageBox

**Comment by cas--** [comment]
I think adding this to onApply should suffice:
```
            if ('language' in changed) {
                Ext.Msg.alert(_('Interface'), _('Please refresh browser...'));
            }
```

File: deluge/ui/web/js/deluge-all/preferences/InterfacePage.js [hunk 5]
```diff
@@ -13,16 +13,22 @@ Ext.namespace('Deluge.preferences');
...
+            width: 320
+        }, config);
+        Deluge.preferences.Interface.superclass.constructor.call(this, config);
+    },
 
     initComponent: function() {
         Deluge.preferences.Interface.superclass.initComponent.call(this);
 
-        var om = this.optionsManager = new Deluge.OptionsManager();
+        var om = this.optionsManager = deluge.preferences.getOptionsManager();
```
Comment (bendikro) [inline deluge/ui/web/js/deluge-all/preferences/InterfacePage.js]: This causes an error when trying to save when not connected to a daemon. It should be possible to save the Interface page when not connected.

Comment (cas--) [inline deluge/ui/web/js/deluge-all/preferences/InterfacePage.js]: This change causes an error?

Comment (cas--) [inline deluge/ui/web/js/deluge-all/preferences/InterfacePage.js]: I shall revert it

Comment (bendikro) [inline deluge/ui/web/js/deluge-all/preferences/InterfacePage.js]: Reverted this change and added the refresh popup

**Pull request merged**